text/0000-machine-input.md

```diff
@@ -0,0 +1,222 @@
+
+**Machine factory**
+
+```js
+const createGreetMachine = (name) => {
+  return createMachine({
+    entry: (ctx) => { console.log(`Hello, ${ctx.name}!`); }
+  });
+
+  // no guarantee that a different createMachine(...) won't be returned
+  // depending on arguments
+}
+
+interpret(createGreetMachine('David')).start();
+```
+
+Problems:
+- Machine definition is oblivious to any input properties
+- Machine is dynamically created; cannot guarantee that the machine is the same shape every time (even though it should always be)
+- Machine definition cannot be serialized; must be run ad-hoc rather than being directly interpreted
+- Machine is no longer easily portable between different languages (future goal)
+- No standard way of creating a machine factory
+
+**Bespoke input event**
+
+```js
+const greetMachine = createMachine({
+  on: {
+    'GREET.INPUT': {
+      actions: (_, event) => { console.log(`Hello, ${event.name}!`); }
+    }
+  }
+});
+
+const greetService = interpret(greetMachine).start();
+
+greetService.send({ type: 'GREET.INPUT', name: 'David' });
+```
+
+Problems:
+- Since this input event is bespoke, there is no built-in way of identifying input properties
+- Input event _must_ be first event sent; this is not guaranteed
+- **Cannot use input data at semantic start of machine; e.g., not usable in `entry`**
+- No standard input event shape
+
+-----
+
+**Prior art:**
+
+- [Data in AWS Step Functions (States Language)](https://states-language.net/spec.html#data):
+
+> The interpreter passes data between states to perform calculations or to dynamically control the state machine’s flow. All such data MUST be expressed in JSON.
+>
+> When a state machine is started, the caller can provide an initial JSON text as input, which is passed to the machine's start state as input. If no input is provided, the default is an empty JSON object, `{}`. As each state is executed, it receives a JSON text as input and can produce arbitrary output, which MUST be a JSON text. [...]
+
+- [Azure Logic Apps](https://docs.microsoft.com/en-us/azure/logic-apps/logic-apps-overview) have _triggers_, which are events (with input data) that start a logic apps workflow.
+- [Google Cloud Platform Workflows](https://cloud.google.com/workflows/docs/passing-runtime-arguments) enable `params` that let you specify data in a workflow execution request
+- [Pipedream](https://pipedream.com/docs/workflows/steps/triggers/#app-based-triggers) has triggers that can contain input data for the start of a workflow.
+- [Argo workflow inputs](https://argoproj.github.io/argo-workflows/workflow-inputs/)
+- [GitHub Actions inputs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs)
+
+## Detailed design
+
+**Overall API**
+
+This RFC proposes to extend `{ type: 'xstate.init' }`, which is always the first event that a machine receives, with the optional `data?: ...` property:
```

Comment:
We can remove the optional parts of this line if we agree on providing an empty object as default.